libs/cardano-ledger-core/src/Cardano/Ledger/UMap.hs

```diff
@@ -271,539 +398,578 @@ toUMapPair (UMap !m1 !m2) =
+-- This is only implemented and is applicable to `RewDepUView`s.
+--
+-- We presume that the domain of the `Map` on the right, is a subset of the domain of the `UView` on the left.
+--
+-- Spec:
 -- evalUnified (rewards dState ∪+ registeredAggregated)
 -- evalUnified (rewards' ∪+ update)
--- evalUnified  (RewardDeposits u0 ∪+ refunds)
-
--- | Add the aggegated Coin from aggRewMap to the RewardDeposits view of the UMap
---   we assume the domain of aggRewMap is a subset of the domain of 'tripmap'
---   The other Views (other than RewardDeposits) are not reachable, since they do not have DPair as 3rd parameter
-(∪+) ::
-  View c k RDPair ->
-  Map k (CompactForm Coin) ->
-  UMap c
-(RewardDeposits (UMap tripmap ptrmap)) ∪+ aggRewMap = UMap (unionHelp tripmap aggRewMap) ptrmap
-
-addCoinToJustRewardsPartOfRDPair :: StrictMaybe RDPair -> CompactForm Coin -> StrictMaybe RDPair
-addCoinToJustRewardsPartOfRDPair SNothing _ = SNothing
-addCoinToJustRewardsPartOfRDPair (SJust (RDPair rew deposit)) delta =
-  SJust (RDPair (addCompact rew delta) deposit)
-
-unionHelp ::
-  Ord k =>
-  Map k (Trip c) ->
-  Map k (CompactForm Coin) ->
-  Map k (Trip c)
-unionHelp tm mm =
-  let f _k (Triple p1 s deposit) delta =
-        Just (Triple (addCoinToJustRewardsPartOfRDPair p1 delta) s deposit)
-      -- We use Map.empty below because mm is a subset of tm, we never add anything here.
-      result = Map.mergeWithKey f id (const Map.empty) tm mm
-   in assert (Map.valid result) result
-
-unionKeyDeposits :: View c k RDPair -> Map k (CompactForm Coin) -> UMap c
-unionKeyDeposits view coinmap = unView (Map.foldlWithKey' accum view coinmap)
+-- evalUnified (RewDepUView u0 ∪+ refunds)
+unionRewAgg, (∪+) :: UView c k RDPair -> Map k (CompactForm Coin) -> UMap c
+(RewDepUView UMap {..}) ∪+ aggRewMap = UMap newUmElem umPtr
   where
-    accum view1 k ccoin = insertWith' combine k (RDPair (CompactCoin 0) ccoin) view1
-    combine (RDPair rew dep) (RDPair _ delta) = RDPair rew (addCompact dep delta)
-
--- ============================================
--- evalUnified (setSingleton hk ⋪ RewardDeposits u0)
--- evalUnified (setSingleton hk ⋪ Delegations u1)
-
-(⋪) ::
-  Set k ->
-  View c k v ->
-  UMap c
-set ⋪ view = unView (Set.foldl' (flip delete') view set)
+    newUmElem =
+      let
+        result = Map.mergeWithKey f id (const Map.empty) umElem aggRewMap
+        f _k (UElemP p1 s deposit drep) delta = Just (UElemP (addC delta p1) s deposit drep)
+       in
+        -- We use Map.empty below because aggRewMap is a subset of umElem, we never add anything here.
+
+        assert (Map.valid result) result
+    addC :: CompactForm Coin -> StrictMaybe RDPair -> StrictMaybe RDPair
+    addC newR = \case
+      SNothing -> SNothing
+      SJust (RDPair r d) -> SJust $ RDPair (addCompact r newR) d
+unionRewAgg = (∪+)
+
+-- | QUESTION: Why is this not the correct implementation for `unionRewAgg`?
```

Comment:
@TimSheard @lehins I have this question about the implementation. If `unionKeyRewards` can replace `unionRewAgg` I will do so.

Reply:
@TimSheard I have updated the `CHANGELOG` as well as retained a commit in the history that demonstrates the property test based equivalence between the older `unionRewAgg` and the new `unionKeyRewards`.

I was also able to remove the `assert (Map.valid result) result` as part of this. cc @lehins @JaredCorduan